libftl/ingest.c

```diff
@@ -92,6 +92,96 @@ OS_THREAD_ROUTINE _ingest_get_rtt(void *data) {
     return 0;
 }
 
+ftl_status_t find_closest_available_ingest(const char* ingestHosts[], int ingestsCount, char* bestIngestHostComputed)
+{
+    ftl_ingest_t* ingestElements;
+    int i;
+
+    if ((ingestElements = malloc(sizeof(ftl_ingest_t) * ingestsCount)) == NULL) {
+        return FTL_MALLOC_FAILURE;
+    }
+
+    for (i = 0; i < ingestsCount; i++) {
+        size_t host_len = strlen(ingestHosts[i]);
+        if ((ingestElements[i].hostname = malloc(host_len)) == NULL) {
+            free(ingestElements);
+            return FTL_MALLOC_FAILURE;
+        }
+        strcpy_s(ingestElements[i].hostname, host_len, ingestHosts[i]);
+        ingestElements[i].rtt = 1000;
+        ingestElements[i].next = NULL;
+    }
+
+    OS_THREAD_HANDLE *handles;
+    _tmp_ingest_thread_data_t *data;
+
+    ftl_ingest_t *elmt, *best = NULL;
+    struct timeval start, stop, delta;
+
+    if ((handles = (OS_THREAD_HANDLE *)malloc(sizeof(OS_THREAD_HANDLE) * ingestsCount)) == NULL) {
+        free(ingestElements);

```

Comment:
This doesn't free all of the memory allocated in the for loop for the strings.

Reply:
A common pattern to use here is to wrap the logic in a do{} while(0); Then as you allocate memory you can keep running in the do{}, if you fail out you can break and do all of your cleanup code at the end of the while.